### PR TITLE
Introduce GameProfileCache

### PR DESCRIPTION
**Pull request by kashike**
[API](https://github.com/SpongePowered/SpongeAPI/pull/1068) | [**Common**](https://github.com/SpongePowered/SpongeCommon/pull/495)

### CODE REVIEW

File: src/main/java/org/spongepowered/common/mixin/core/server/management/MixinPlayerProfileCache.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+                result.add((GameProfile) profile);
+            }
+        }
+
+        return result.isEmpty() ? ImmutableList.of() : ImmutableList.copyOf(result);
+    }
+
+    @Override
+    public Collection<GameProfile> lookupBulkId(Iterable<UUID> uniqueIds) {
+        checkNotNull(uniqueIds, "unique ids");
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/common/mixin/core/server/management/MixinPlayerProfileCache.java]: As on the API PR, should be `uniqueIds`

File: src/main/java/org/spongepowered/common/mixin/core/server/management/MixinPlayerProfileCache.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        }
+
+        this.addEntry((com.mojang.authlib.GameProfile) profile, expiry);
+
+        return true;
+    }
+
+    @Override
+    public Collection<GameProfile> getBulkId(Iterable<UUID> uniqueIds) {
+        checkNotNull(uniqueIds, "unique ids");
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/common/mixin/core/server/management/MixinPlayerProfileCache.java]: `uniqueIds`

File: src/main/java/org/spongepowered/common/mixin/core/server/management/MixinPlayerProfileCache.java [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+            // the callback here.
+            callback.onProfileLookupSucceeded((com.mojang.authlib.GameProfile) cache.getOrLookupByName(names[0]).orElse(null));
+        }
+    }
+
+    @Nullable
+    public com.mojang.authlib.GameProfile getByNameNoLookup(String username) {
+        @Nullable IMixinPlayerProfileCacheEntry entry = this.usernameToProfileEntryMap.get(username.toLowerCase(Locale.ROOT));
+
+        if (entry != null && new Date().getTime() >= entry.getExpirationDate().getTime()) {
```
Comment (jk-5) [inline src/main/java/org/spongepowered/common/mixin/core/server/management/MixinPlayerProfileCache.java]: You should use `System.currentTimeMillis()`. Eliminates 1 object creation and does the same (not that 1 object allocation is really bad, but might as well be clean)

Comment (kashike) [inline src/main/java/org/spongepowered/common/mixin/core/server/management/MixinPlayerProfileCache.java]: `new Date().getTime() >= entry.getExpirationDate().getTime()` comes from `PlayerProfileCache`, from vanilla. I'll change this soonish - thanks!

**Comment by gabizou** [comment]
Merged in https://github.com/SpongePowered/SpongeCommon/commit/c346273b4b3311dd3e4484cbd146c03c28c3f00e

**Pull request closed without merging**